File: AIiOS/AIiOS/Services/UsernameValidator.swift
import Foundation

enum UsernameValidator {
    // 4–16 chars, lowercase letters / digits / underscore only
    private static let regex = try! NSRegularExpression(pattern: "^[a-z0-9_]{4,16}$")

    enum ValidationError: LocalizedError {
        case tooShort, tooLong, invalidChars
        var errorDescription: String? {
            switch self {
            case .tooShort:     return "Username must be at least 4 characters."
            case .tooLong:      return "Username must be 16 characters or fewer."
            case .invalidChars: return "Only lowercase letters, numbers, and underscores."
            }
        }
    }

    static func validate(_ username: String) throws {
        if username.count < 4  { throw ValidationError.tooShort }
        if username.count > 16 { throw ValidationError.tooLong }
        let range = NSRange(username.startIndex..., in: username)
        if regex.firstMatch(in: username, range: range) == nil {
            throw ValidationError.invalidChars
        }
    }

    static func normalized(_ raw: String) -> String {
        raw.lowercased().trimmingCharacters(in: .whitespaces)
    }
}

File: AIAndroid/app/src/main/kotlin/ai/ecoinference/app/services/UsernameValidator.kt
package ai.ecoinference.app.services

object UsernameValidator {
    private val regex = Regex("^[a-z0-9_]{4,16}$")

    sealed class ValidationError(msg: String) : Exception(msg) {
        object TooShort     : ValidationError("Username must be at least 4 characters.")
        object TooLong      : ValidationError("Username must be 16 characters or fewer.")
        object InvalidChars : ValidationError("Only lowercase letters, numbers, and underscores.")
    }

    fun validate(username: String) {
        if (username.length < 4)  throw ValidationError.TooShort
        if (username.length > 16) throw ValidationError.TooLong
        if (!regex.matches(username)) throw ValidationError.InvalidChars
    }

    fun normalized(raw: String): String = raw.lowercase().trim()
}
